### ISSUE TITLE
Composer environment with service/pods IP Aliasing using provided subnet secondary range

### ISSUE THREAD

**Issue by tomhipwell** [terraform-providers/terraform-provider-google#2875]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment
* If an issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to "hashibot", a community member has claimed the issue already.

### Terraform Version

Terraform v0.11.11
+ provider.google v1.19.1
+ provider.google-beta v1.20.0

### Affected Resource(s)

google_composer_environment

### Terraform Configuration Files

Main composer module, var.subnetwork and var.network are self links to the relevant resources. I'm using the beta provider for this resource (I've abbreviated it slightly to remove the software config block):

```tf
provider "google-beta" {
  version = "~> 1.19"
  region  = "${var.region}"
  project = "${var.project_id}"
}

resource "google_composer_environment" "airflow" {
  provider = "google-beta"
  name     = "${var.project_id}-airflow"
  project  = "${var.project_id}"
  region   = "${var.region}"

  config {
    node_count = "${var.primary_nodes}"

    node_config {
      zone            = "${var.zone}"
      machine_type    = "n1-standard-1"
      network         = "${var.network}"
      subnetwork      = "${var.subnetwork}"
      service_account = "${var.composer_service_account}"
    }
  }
}
```
The subnet definition, in a separate module (the self-link of this resource is passed to the composer module). The subnet sits on a shared vpc network, though the composer environment sits in the host project, not a service project:
```tf
resource "google_compute_subnetwork" "composer_subnet" {
  name          = "${var.project_id}-${var.composer_region}-composer-subnet"
  project       = "${var.project_id}"
  network       = "${google_compute_network.vpc_network.self_link}"
  ip_cidr_range = "10.${var.subnet_secondary_range_root}4.0.0/16"
  region        = "${var.composer_region}"

  secondary_ip_range {
    range_name    = "composer-pods"
    ip_cidr_range = "10.${var.subnet_secondary_range_root}5.0.0/21"
  }

  secondary_ip_range {
    range_name    = "composer-services"
    ip_cidr_range = "10.${var.subnet_secondary_range_root}6.0.0/21"
  }
}
```
### Expected Behavior

The cloud composer environment was built with the underlying GKE cluster having ip aliasing enabled and pods/services sat on the supplied secondary ranges (composer-pods, composer-services).

### Actual Behavior

IP aliasing disabled, services/pods do not use the provided secondary ranges of the subnet.

### Notes

I've followed the composer guide to shared vpc in my network design:
https://cloud.google.com/composer/docs/how-to/managing/configuring-shared-vpc

The docs imply this should work with the beta provider:
https://www.terraform.io/docs/providers/google/r/composer_environment.html

The api docs imply that it just needs the network and subnetwork self-links, and the subnetwork needs to be correctly configured with the right secondary ranges, which it is - so the right arguments are being collected and passed to the resource:
https://developers.google.com/resources/api-libraries/documentation/composer/v1beta1/python/latest/composer_v1beta1.projects.locations.environments.html

I have a separate GKE cluster deployed in the same project using the google_container_resource but the same pattern of subnetwork design (secondary ranges are specified via the ip_allocation_policy config block) which is working fine.

**Comment by emilymye** [comment]
Sorry in advance, I'm not entirely familar with setting up shared VPC for Composer or GKE, so please bear with me! 

**tl;dr:** please try the following things again for me:
* Put the Composer environment in a separate service project (i.e. not the host project)
* The same thing but with debug logs on (`TF_LOG="DEBUG"`)

The only possible thing I can see that's different from the [docs](https://cloud.google.com/composer/docs/how-to/managing/configuring-shared-vpc) is that it looks like they expect the composer environment to belong to the service project:

> Shared VPC requires that you designate a host project to which networks and subnetworks belong and a service project, which is attached to the host project. When Cloud Composer participates in a Shared VPC, **the Cloud Composer environment is in the service project**.

and a project can't be a host and service project simultaneously (From https://cloud.google.com/vpc/docs/shared-vpc). I'm not sure why the GKE cluster would work in the same project. Have you tried using a separate service project? 

Otherwise, based on the fact that we don't do anything in particular to enable Shared VPC on the Terraform side, there are really only a few things happening here that I can think of (by IMO-likelihood) : 
1. Something is missing from your resources, configuration, or GCP project state (e.g. IAM permissions)
2. The Composer documentation is wrong and something else needs to be configured (goes in hand with 1)
3. The Composer set-up of the GKE cluster is wrong in a way that prevents proper shared VPC set-up.
4. Terraform isn't sending something to the Composer API.

I would love to help with any of these issues, but to do so, I'd need to know what the actual GCP resources look like. Would you mind posting your debug logs (i.e. run with `TF_LOG="DEBUG"`) so we can see what's being sent to/returned from the API? 






I'm not sure I actually

**Comment by bjaworski3** [comment]
I managed to get this working with `provider.google-beta v2.1.0`

As mentioned by @emilymye you will need two separate projects, a host project that contains the VPCs, and a service project that is attached to the host project that will contain the Composer/GKE Cluster.

I didn't include it in the terraform below, but there are also a couple IAM roles that need to be set on the service-project.

From https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-shared-vpc :

The `service-[SERVICE_PROJECT_1_NUM]@container-engine-robot.iam.gserviceaccount.com` service account is granted `roles/compute.networkUser` in the host project
The `[SERVICE_PROJECT_1_NUM]@cloudservices.gserviceaccount.com` service account is granted  `roles/compute.networkUser` in the host project
The `service-[SERVICE_PROJECT_1_NUM]@container-engine-robot.iam.gserviceaccount.com` is granted `roles/container.hostServiceAgentUser` in the host project

Here's an example config:

```
resource "google_compute_network" "network" {
  project                 = "host-project"
  name                    = "composer-network"
  auto_create_subnetworks = false
}

resource "google_compute_subnetwork" "composer" {
  project                  = "host-project"
  network                  = "composer-network"
  name                     = "composer-subnetwork"
  region                   = "us-east4"
  private_ip_google_access = true
  ip_cidr_range            = "10.10.0.0/16"
  secondary_ip_range       = [
    {
      range_name    = "composer-pods"
      ip_cidr_range = "10.11.0.0/16"
    },
    {
      range_name    = "composer-services"
      ip_cidr_range = "10.12.0.0/16"
    },
  ]
}

resource "google_compute_shared_vpc_host_project" "host" {
  project = "host-project"
}

resource "google_compute_shared_vpc_service_project" "service1" {
  host_project    = "${google_compute_shared_vpc_host_project.host.project}"
  service_project = "service-project"
}


resource "google_composer_environment" "composer" {
  provider = "google-beta"
  name     = "shared-vpc"
  project  = "service-project"
  region   = "us-east4"

  config {
    node_count = "3"

    node_config {
      zone            = "us-east4-a"
      machine_type    = "n1-standard-1"
      network         = "${google_compute_network.network.self-link}"
      subnetwork      = "${google_compute_subnetwork.composer.self-link}"
    }
  }
}
```

**Comment by tomhipwell** [comment]
Awesome - thanks, will give it a whirl!
